contracts/core/Allo.sol

```diff
@@ -33,57 +57,108 @@ contract Allo is Initializable {
+    /// @dev invoked only by pool owner
+    function updatePoolMetadata(
         uint _poolId,
-        bytes memory _data
+        bytes memory _metadata
     ) external payable returns (bytes memory) {
-        // Implement the function here
+        if (!registry.isMemberOfIdentity(_identityId, msg.sender)) {
+            revert NO_ACCESS_TO_ROLE();
+        }
+
+        Pool storage pool = pools[_poolId];
+        pool.metadata = _metadata;
+
+        emit PoolMetadataUpdated(_poolId, _metadata);
     }
 
     // transfers _poolAmt from msg.sender to the pool, and takes a fee
     function fundPool(uint _poolId, uint _poolAmt) external payable {
         // Implement the function here
     }
 
-    // passes _data & msg.sender through to the allocation strategy for that pool
+    /// @notice passes _data & msg.sender through to the allocation strategy for that pool
+    /// @param _poolId id of the pool
+    /// @param _data encoded data unique to the allocation strategy for that pool
     function allocate(uint _poolId, bytes memory _data) external payable {
-        // Implement the function here
+        pools[_poolId].allocationStrategy.allocate(_data, msg.sender);
     }
 
     // calls voting.generatePayouts() and then uses return data for payout.activatePayouts()
     // permissionless for anyone to call, checks happen within the strategies
     // check to make sure they haven't skrited around fee
-    function finalize(uint _poolId) external {
-        // Implement the function here
+    function finalize(uint _poolId, bytes memory _dataFromSender) external {
+
+        // ASK: Do we need _data to allow owner to pass custom data ?
+        if (!registry.isMemberOfIdentity(_identityId, msg.sender)) {
+            revert NO_ACCESS_TO_ROLE();
+        }
+
+        Pool memory pool = pools[_poolId];
+        bytes memory dataFromAllocationStrategy = pool.allocationStrategy.generatePayouts();
+        pool.disributionStrategy.activatePayouts(dataFromAllocationStrategy, _dataFromSender);
+
     }
 
-    // call to payoutStrategy.distribute()
+    /// @notice passes _data & msg.sender through to the disribution strategy for that pool
+    /// @param _poolId id of the pool
+    /// @param _data encoded data unique to the disributionStrategy strategy for that pool
     function distribute(uint _poolId, bytes memory _data) external {
-        // Implement the function here
+        if (!registry.isMemberOfIdentity(_identityId, msg.sender)) {
+            revert NO_ACCESS_TO_ROLE();
+        }
+        pools[_poolId].disributionStrategy.distribute(_data, msg.sender);
     }
 
-    // call to distributionStrategy.close()
+    /// @notice Closes the pool
     function closePool(uint _poolId) external {
-        // Implement the function here
+        // pools[_poolId].allocationStrategy.close(); // ASK: Do we need this ?
+        // pools[_poolId].distributionStrategy.close(); // ASK: Do we need this ?
+        pools[_poolId].active = false;
```

Comment:
@zobront our distributionStrategy interface doesn't have a close function.
Would it make sense to add it do both allocationStrategy and distributionStrategy ?

Reply:
remove `closePool` and have active derived from strategies